Security fixes

Rytautas opened the pull request.

mildapetr committed:
fix/registration_errors (#5)

* fixed id and error issues

* simplifying update by not allowing to update values (except for admin review status later)

* fix/minor changes

* minor changes #2

Co-authored-by: Rytautas <rytautas.kvasinskas@gmail.com>

mildapetr committed:
test/user_validation (#6)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

Rytautas committed:
added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

Rytautas committed:
added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

Rytautas committed:
added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

Rytautas committed:
added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

Rytautas committed:
changed username to email as we are using email instead of username

Rytautas committed:
changed username to email as we are using email instead of username

Rytautas committed:
fixed bean autowiring issue

Rytautas committed:
fixed bean autowiring issue

Rytautas committed:
fixed issue with POST request for /auth/login being denied. it now returns an internal server error
minor changes to DB that will need to be fixed later

Rytautas committed:
Merge remote-tracking branch 'origin/security' into security

# Conflicts:
#	src/main/java/com/project/security/SecurityConfig.java
#	src/main/resources/db/migration/V1.0__init.sql

Rytautas committed:
changed autowiring to go on constructor and not on field

Rytautas committed:
some changes

Rytautas committed:
change session duration to 15min

GaiJinn committed:
fix: fix user authentication and password encryption

Rytautas committed:
fixed REST controllers returning too much information when creating and updating users/applications

Rytautas committed:
Security (#9)

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* fixed issue with POST request for /auth/login being denied. it now returns an internal server error
minor changes to DB that will need to be fixed later

* changed autowiring to go on constructor and not on field

* some changes

* change session duration to 15min

* fix/registration_errors (#5)

* fixed id and error issues

* simplifying update by not allowing to update values (except for admin review status later)

* fix/minor changes

* minor changes #2

Co-authored-by: Rytautas <rytautas.kvasinskas@gmail.com>

* test/user_validation (#6)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* fix: fix user authentication and password encryption

* fixed REST controllers returning too much information when creating and updating users/applications

Co-authored-by: mildapetr <58117403+mildapetr@users.noreply.github.com>
Co-authored-by: gaijinn <gaijinn@users.noreply.github.com>

Rytautas committed:
database changes for the upcoming features (#8)

Rytautas committed:
formatted code

Rytautas committed:
removed unnecessary imports

Rytautas committed:
fixed warnings with ResponseEntity usage

Rytautas committed:
fixed issues that appeared after merging security branch
made some changes to the database

Rytautas committed:
Merge branch 'master' into security-fixes

Rytautas committed:
added encrypted passwords to database

Rytautas committed:
Merge remote-tracking branch 'origin/security-fixes' into security-fixes

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -25,29 +24,29 @@ public UserController(UserService userService) {
     @PostMapping
-    ResponseEntity createUser(@RequestBody User newUser) throws UserException {
+    ResponseEntity<HttpStatus> createUser(@RequestBody User newUser) throws UserException {
         User createdUser = userService.addUser(newUser);
```
This `createdUser` is never used.

Rytautas replied:
removed the unnecessary assignment.
made corresponding changes to ApplicationController class

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -25,29 +24,29 @@ public UserController(UserService userService) {
 
     @GetMapping
-    public ResponseEntity<List<User>> getUsers(){
+    public ResponseEntity<List<User>> getUsers() {
```
If you have `fetchUser` then this should be called `fetchUsers`.

Rytautas replied:
renamed getUsers to fetchAllUsers.
made corresponding changes to ApplicationController class

grkarolis commented on src/main/java/com/project/controller/ApplicationController.java:
```diff
@@ -35,18 +35,18 @@ public ApplicationController(ApplicationService applicationService) {
 
     @PostMapping
-    ResponseEntity createApplication(@Valid @RequestBody Application newApplication) throws UserException, ApplicationNotFoundException {
+    ResponseEntity<HttpStatus> createApplication(@Valid @RequestBody Application newApplication) throws UserException, ApplicationNotFoundException {
```
No need to say it's a `newApplication` calling it just `application` is good enough.

Rytautas replied:
renamed newApplication to application
made corresponding changes to UserController class

grkarolis commented on src/main/java/com/project/DemoApplication.java:
```diff
@@ -11,8 +11,9 @@
-		SpringApplication.run(DemoApplication.class, args);
-	}
+    public static void main(String[] args) {
+        SpringApplication.run(DemoApplication.class, args);
```
I highly doubt that your application is called `DemoApplication`.

Rytautas replied:
renamed DemoApplication class to ITAcademyBackend

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -21,53 +22,53 @@
     @Size(max = 20, message = "Telefono numeris negali būti ilgesnis nei 20 simbolių")
-    @Column(name="phone_number")
+    @Column(name = "phone_number")
     private String phone_number;
```
Please change variable name to `phoneNumber`.

Rytautas replied:
changed the names of all variables in Application class to camel case

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -21,53 +22,53 @@
     @Size(max = 255, message = "Vardas negali būti ilgesnis nei 255 simboliai")
-    @Column(name="full_name")
+    @Column(name = "full_name")
     private String name;
```
If this is a full name then your variable should reflect that. It should not be `name`. Should be `fullName`.

Rytautas replied:
changed several fields' names in Application class to better reflect what values are held in them

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -21,53 +22,53 @@
     @Size(max = 1500, message = "Pomėgių lauke negali būti daugiau nei 1500 simbolių")
-    @Column(name="free_time")
+    @Column(name = "free_time")
     private String free_time;
```
Maybe this could be not `free_time` but `hobbies`? P.S Never user underscore when naming your variables. Unless they are static final constants.

Rytautas replied:
changed several fields' names in Application class to better reflect what values are held in them

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -21,53 +22,53 @@
     @NotNull(message = "Ar tinka akademijos laikas privalo būti pasirinkta")
-    @Column(name="academy_time")
+    @Column(name = "academy_time")
     private boolean academy_time;
```
Maybe `isAcademyTimeSuitable` would be a better name. Since this is a logic variable. It's hard to understand that does `academy_time` mean.

Rytautas replied:
changed several fields' names in Application class to better reflect what values are held in them

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -21,53 +22,53 @@
     @NotBlank(message = "Paraiškos data negali būti tuščia")
-    @Column(name="application_date")
+    @Column(name = "application_date")
     private String application_date;
```
As mentioned above. Don't use underscore when naming your variables. Change it to something like `applicationDate`.

Rytautas replied:
changed several fields' names in Application class to better reflect what values are held in them

grkarolis commented on src/main/java/com/project/security/SecurityConfig.java:
```diff
@@ -61,9 +60,9 @@ protected void configure(HttpSecurity http) throws Exception {
                 .antMatchers(HttpMethod.GET, "/api/applications").authenticated()
 
                 //.anyRequest().authenticated()
```
Is this necessary?

Rytautas replied:
removed unnecessary comment

grkarolis commented on src/main/java/com/project/service/ApplicationServiceImpl.java:
```diff
@@ -41,14 +40,14 @@ public Application addApplication(Application application) throws UserException
     @Override
     public void deleteApplication(long id) throws ApplicationNotFoundException {
-        if(!applicationRepository.findById(id).isPresent())
+        if (!applicationRepository.findById(id).isPresent())
```
`if` statements should be opened and closed with curly braces `{}` even if they have only one line of code.

Rytautas replied:
added curly braces {} for if statements in UserServiceImpl and ApplicationServiceImpl classes

grkarolis commented on src/main/java/com/project/service/ApplicationServiceImpl.java:
```diff
@@ -41,14 +40,14 @@ public Application addApplication(Application application) throws UserException
     @Override
     public Application updateApplication(Application application, long id) throws ApplicationNotFoundException {
-        if(!applicationRepository.findById(id).isPresent())
+        if (!applicationRepository.findById(id).isPresent())
```
`if` statements should be opened and closed with curly braces `{}` even if they have only one line of code.

grkarolis commented on src/main/java/com/project/service/CustomUserDetailsService.java:
```diff
@@ -19,6 +19,6 @@ public CustomUserDetailsService(UserRepository users) {
 
     @Override
     public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {              // email is used as username
```
If this comment is necessary please move it above the `return` statement.

Rytautas replied:
fixed

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -46,24 +45,24 @@ public User addUser(User user) throws UserException {
 
         user.setPassword(bcryptEncoder.encode(user.getPassword()));
-        if(userRepository.findByEmail(user.getEmail()).isPresent())
+        if (userRepository.findByEmail(user.getEmail()).isPresent())
```
`if` statements should be opened and closed with curly braces `{}` even if they have only one line of code.

Rytautas replied:
added curly braces `{}` for `if` statements in UserServiceImpl and ApplicationServiceImpl classes

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -46,24 +45,24 @@ public User addUser(User user) throws UserException {
     @Override
     public User updateUser(User user, long id) throws UserException {
-        if(!userRepository.findById(id).isPresent())
+        if (!userRepository.findById(id).isPresent())
```
`if` statements should be opened and closed with curly braces `{}` even if they have only one line of code.

Rytautas replied:
added curly braces {} for if statements in UserServiceImpl and ApplicationServiceImpl classes

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -46,24 +45,24 @@ public User addUser(User user) throws UserException {
         userValidator.validate(user);
 
-        if(userRepository.findByEmail(user.getEmail()).isPresent()
+        if (userRepository.findByEmail(user.getEmail()).isPresent()
```
`if` statements should be opened and closed with curly braces `{}` even if they have only one line of code.

Rytautas replied:
added curly braces {} for if statements in UserServiceImpl and ApplicationServiceImpl classes

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -79,7 +78,7 @@ public User updateUser(User user, long id) throws UserException {
     @Override
     public void deleteUser(long id) throws UserNotFoundException {
-        if(!userRepository.findById(id).isPresent())
+        if (!userRepository.findById(id).isPresent())
```
`if` statements should be opened and closed with curly braces `{}` even if they have only one line of code.

Rytautas replied:
added curly braces `{}`  to all `if` statements in UserServiceImpl and ApplicationServiceImpl classes

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -46,24 +45,24 @@ public User addUser(User user) throws UserException {
         List<String> roles = new ArrayList<>();
-        roles.add("User");
+        roles.add("USER");
         user.setRoles(roles); //probably more logical as we don't have admin registration
```
Don't think this comment should be here.

Rytautas replied:
removed unnecessary comments

grkarolis commented on src/main/java/com/project/model/User.java:
```diff
@@ -1,20 +1,25 @@
 import org.springframework.security.core.authority.SimpleGrantedAuthority;
 import org.springframework.security.core.userdetails.UserDetails;
 
 import javax.persistence.*;
```
You should not import everything from a package if you do not need it. Better to just import what you need.

Rytautas replied:
Intellij automatically changes imports to import package.* when there are more than 4 elements imported from the same package. As there are some classes with more than 8 or 10 imports from the same package, i think it is better to let intellij handle the imports as writing them manually every time you add a new class to import from the same package is going to be time consuming

grkarolis commented on src/main/java/com/project/model/User.java:
```diff
@@ -1,20 +1,25 @@
-import java.util.Collection;
-import java.util.List;
+
+import java.util.*;
```
You should not import everything from a package if you do not need it. Better to just import what you need.

Rytautas replied:
Intellij automatically changes imports to import package.* when there are more than 4 elements imported from the same package. As there are some classes with more than 8 or 10 imports from the same package, i think it is better to let intellij handle the imports as writing them manually every time you add a new class to import from the same package is going to be time consuming

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -4,6 +4,7 @@
 import lombok.NoArgsConstructor;
 import lombok.ToString;
+
 import javax.persistence.*;
```
You should not import everything from a package if you do not need it. Better to just import what you need.

Rytautas replied:
Intellij automatically changes imports to import package.* when there are more than 4 elements imported from the same package. As there are some classes with more than 8 or 10 imports from the same package, i think it is better to let intellij handle the imports as writing them manually every time you add a new class to import from the same package is going to be time consuming

Rytautas committed:
removed the unnecessary assignment in UserController class
renamed getUsers to fetchAllUsers in UserController class
renamed newUser to user in UserController class
made corresponding changes to ApplicationController class
added curly braces {} for if statements in UserServiceImpl and ApplicationServiceImpl classes

renamed DemoApplication class to ITAcademyBackend
changed the names of all variables in Application class to camel case
changed several fields' names in Application class to better reflect what values are held in them
removed unnecessary comments

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -25,29 +24,30 @@ public UserController(UserService userService) {
 
     @GetMapping
-    public ResponseEntity<List<User>> getUsers(){
+    public ResponseEntity<List<User>> fetchAllUsers() {
```
You don't need to explicitly say that it will fetch `all` users. It is already clear from you `Get` since it has no request parameters.

grkarolis commented on src/main/java/com/project/controller/ApplicationController.java:
```diff
@@ -24,7 +24,7 @@ public ApplicationController(ApplicationService applicationService) {
 
     @GetMapping
-    public ResponseEntity<List<Application>> getAll() {
+    public ResponseEntity<List<Application>> fetchAllApplications() {
```
You don't need to explicitly say that it will fetch `all` applications. It is already clear from you `Get` since it has no request parameters.

grkarolis commented on src/main/java/com/project/security/JwtTokenProvider.java:
```diff
@@ -67,7 +68,7 @@ public String getUserEmail(String token) {
+        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
             return bearerToken.substring(7);
         }
         return null;
```
If you need to return null maybe you could consider using `Optional<>`.

grkarolis reviewed (approved):
Good job!
